### ISSUE TITLE
Rare errors on alcOpenDevice or alcCreateContext with Oboe backend.

### ISSUE THREAD

**Issue by efroemling** [kcat/openal-soft#950]
Once more, thanks for all the help getting things polished up with my OpenSL->Oboe backend transition. The logs I'm seeing are looking pretty clear at this point aside from this one last issue, so there's light at the end of the tunnel.

The first 2 calls I make in my implementation are `alcOpenDevice(nullptr)` followed by `alcCreateContext(device, nullptr)` and I'm seeing a small but substantial number of devices failing on one call or the other (much more than I saw with the OpenSL backend). It seems that failing on `alcCreateContext()` is more common but I'm seeing cases of both.

I just added code to capture logs in these cases and am seeing the first results for the `alcOpenDevice` fail cases:

```
------------------------ OPENALSOFT-FATAL-ERROR-LOG-BEGIN ----------------------
openal-log(0.506624s): [ALSOFT] (II) Initializing library v1.23.1-1381a951 HEAD
openal-log(0.506683s): [ALSOFT] (II) Supported backends: oboe, null
openal-log(0.506690s): [ALSOFT] (II) Loading config /etc/openal/alsoft.conf...
openal-log(0.506953s): [ALSOFT] (II) Loading config /etc/xdg/alsoft.conf...
openal-log(0.507086s): [ALSOFT] (II) Got binary: "/system/bin", "app_process64"
openal-log(0.507098s): [ALSOFT] (II) Loading config /system/bin/alsoft.conf...
openal-log(0.507146s): [ALSOFT] (II) Extensions: +NEON
openal-log(0.507182s): [ALSOFT] (II) Initialized backend "oboe"
openal-log(0.507186s): [ALSOFT] (II) Added "oboe" for playback
openal-log(0.507191s): [ALSOFT] (II) Added "oboe" for capture
openal-log(0.507204s): [ALSOFT] (II) Opening default playback device
openal-log(0.515369s): [ALSOFT] (WW) Failed to open playback device: Failed to create stream: ErrorInternal
openal-log(0.515387s): [ALSOFT] (WW) Error generated on device 0x0, code 0xa004
openal-log(0.515391s): [ALSOFT] (II) Freeing device 0xb400006f1d3e4080
------------------------- OPENALSOFT-FATAL-ERROR-LOG-END -----------------------
```
I can post logs for `alcCreateContext` when those start coming in.

These errors seem to be pretty consistent for a given device; I'm getting a large number of logs for a smaller number of devices which seems to imply repeat failed attempts.

Any thoughts on what this implies or what I should do to further diagnose or correct this? I'd be happy to add test log calls or futz with settings or whatever.

**Comment by efroemling** [comment]
And here's the log for cases where alcOpenDevice succeeds but the following alcCreateContext fails:
```
------------------------ OPENALSOFT-FATAL-ERROR-LOG-BEGIN ----------------------
openal-log(0.507826s): [ALSOFT] (II) Initializing library v1.23.1-1381a951 HEAD
openal-log(0.507855s): [ALSOFT] (II) Supported backends: oboe, null
openal-log(0.508526s): [ALSOFT] (II) Loading config /etc/openal/alsoft.conf...
openal-log(0.509062s): [ALSOFT] (II) Loading config /etc/xdg/alsoft.conf...
openal-log(0.509212s): [ALSOFT] (II) Got binary: "/system/bin", "app_process64"
openal-log(0.509220s): [ALSOFT] (II) Loading config /system/bin/alsoft.conf...
openal-log(0.509874s): [ALSOFT] (II) Extensions: +NEON
openal-log(0.510484s): [ALSOFT] (II) Initialized backend "oboe"
openal-log(0.510489s): [ALSOFT] (II) Added "oboe" for playback
openal-log(0.510491s): [ALSOFT] (II) Added "oboe" for capture
openal-log(0.510766s): [ALSOFT] (II) Opening default playback device
openal-log(0.785946s): [ALSOFT] (II) Created device 0xb400006d45158040, "Oboe Default"
openal-log(0.786734s): [ALSOFT] (II) Pre-reset: Stereo, Float32, 48000hz, 960 / 2880 buffer
openal-log(0.799849s): [ALSOFT] (II) Got stream with properties:
StreamID: 0xb400006f1b5face0
DeviceId: 2
Direction: Output
API type: AAudio
BufferCapacity: 5120
BufferSize: 2880
FramesPerBurst: 640
FramesPerDataCallback: 0
SampleRate: 48000
ChannelCount: 2
Format: Float
SharingMode: Shared
PerformanceMode: LowLatency
CurrentState: Open
XRunCount: 0
FramesRead: 0
FramesWritten: 0
openal-log(0.799870s): [ALSOFT] (II) Post-reset: Stereo, Float32, 48000hz, 640 / 2880 buffer
openal-log(0.799892s): [ALSOFT] (II) Stereo rendering
openal-log(0.799898s): [ALSOFT] (II) Channel config, Main: 3, Real: 2
openal-log(0.799902s): [ALSOFT] (II) Allocating 5 channels, 20480 bytes
openal-log(0.799910s): [ALSOFT] (II) Enabling single-band first-order ambisonic decoder
openal-log(0.799917s): [ALSOFT] (II) Max sources: 256 (255 + 1), effect slots: 64, sends: 2
openal-log(0.799921s): [ALSOFT] (II) Dithering disabled
openal-log(0.799924s): [ALSOFT] (II) Output limiter disabled
openal-log(0.799926s): [ALSOFT] (II) Fixed device latency: 0ns
openal-log(0.824638s): [ALSOFT] (EE) Failed to start stream: ErrorDisconnected
openal-log(0.824681s): [ALSOFT] (WW) Error generated on device 0xb400006d45158040, code 0xa001
------------------------- OPENALSOFT-FATAL-ERROR-LOG-END -----------------------
```

**Comment by kcat** [comment]
If it's possible to get an error log from the audio system itself, that may help point to the underlying problem. From OpenAL Soft's perspective it seems the device is just intermittently failing to start or be created. Like something's causing the device to drop out/disconnect/become unavailable right as OpenAL tries to open and start a playback stream (where sometimes it manages to get to the end of context creation when the device is about to start before it's lost, and other times it's right as it's first opened).

Does it work if the process waits a moment before first trying to open the device, or after it first fails, then try again? Is it always at the start of the process, or does it happen after it's been running for a bit too?

**Comment by efroemling** [comment]
I'm running some further tests doing things such as attempting to wait and set up the device/context a second time after a few seconds if it fails and will post any insight I get from that.

Here's a log I just got that might be worth looking at:

In this case it appears that the app was able to create a device and then a context successfully but then almost immediately (1.3 seconds after app launch time) gets a device-disconnected error, which it is able to recover from via a reset. I'm noticing that the stream properties seem to be a bit different after the reset (and more in line with what the backend seems to ask for; ie LowLatency mode). Is that meaningful?
```
openal-log(0.633231s): [ALSOFT] (II) Initializing library v1.23.1-bc83c874 HEAD
openal-log(0.633379s): [ALSOFT] (II) Supported backends: oboe, null
openal-log(0.634039s): [ALSOFT] (II) Loading config /etc/openal/alsoft.conf...
openal-log(0.636378s): [ALSOFT] (II) Loading config /etc/xdg/alsoft.conf...
openal-log(0.636766s): [ALSOFT] (II) Got binary: "/system/bin", "app_process32"
openal-log(0.636794s): [ALSOFT] (II) Loading config /system/bin/alsoft.conf...
openal-log(0.637977s): [ALSOFT] (II) Extensions: +NEON
openal-log(0.638131s): [ALSOFT] (II) Initialized backend "oboe"
openal-log(0.638150s): [ALSOFT] (II) Added "oboe" for playback
openal-log(0.638167s): [ALSOFT] (II) Added "oboe" for capture
openal-log(0.638852s): [ALSOFT] (II) Opening default playback device
openal-log(0.638937s): [ALSOFT] (II) Created device 0x9e256980, "Oboe Default"
openal-log(0.639014s): [ALSOFT] (II) Pre-reset: Stereo, Float32, 48000hz, 960 / 2880 buffer
openal-log(0.666339s): [ALSOFT] (II) Got stream with properties:
StreamID: 0xd2ce2600
DeviceId: 9779
Direction: Output
API type: AAudio
BufferCapacity: 1792
BufferSize: 1792
FramesPerBurst: 882
FramesPerDataCallback: 0
SampleRate: 44100
ChannelCount: 2
Format: Float
SharingMode: Shared
PerformanceMode: None
CurrentState: Open
XRunCount: 0
FramesRead: 0
FramesWritten: 0
openal-log(0.666367s): [ALSOFT] (II) Post-reset: Stereo, Float32, 44100hz, 882 / 1792 buffer
openal-log(0.666392s): [ALSOFT] (II) Stereo rendering
openal-log(0.666551s): [ALSOFT] (II) Channel config, Main: 3, Real: 2
openal-log(0.666561s): [ALSOFT] (II) Allocating 5 channels, 20480 bytes
openal-log(0.667716s): [ALSOFT] (II) Enabling single-band first-order ambisonic decoder
openal-log(0.667791s): [ALSOFT] (II) Max sources: 256 (255 + 1), effect slots: 64, sends: 2
openal-log(0.667905s): [ALSOFT] (II) Dithering disabled
openal-log(0.667926s): [ALSOFT] (II) Output limiter disabled
openal-log(0.667991s): [ALSOFT] (II) Fixed device latency: 0ns
openal-log(0.669387s): [ALSOFT] (II) Post-start: Stereo, Float32, 44100hz, 882 / 1792 buffer
openal-log(0.669695s): [ALSOFT] (II) Increasing allocated voices to 256
openal-log(0.673364s): [ALSOFT] (II) Created context 0x9ee9e480
openal-log(1.301176s): [ALSOFT] (EE) Failed to start stream: ErrorDisconnected
openal-log(1.303814s): [ALSOFT] (WW) Error generated on device 0x9e256980, code 0xa001
DEVICE DISCONNECT DETECTED; ATTEMPTING RESET
openal-log(1.341338s): [ALSOFT] (II) Increasing allocated voices to 256
openal-log(1.357740s): [ALSOFT] (II) Pre-reset: Stereo, Float32, 48000hz, 960 / 2880 buffer
openal-log(1.391200s): [ALSOFT] (II) Error was ErrorDisconnected
openal-log(1.405141s): [ALSOFT] (II) Got stream with properties:
StreamID: 0xd2ce2600
DeviceId: 3
Direction: Output
API type: AAudio
BufferCapacity: 1920
BufferSize: 1920
FramesPerBurst: 240
FramesPerDataCallback: 0
SampleRate: 48000
ChannelCount: 2
Format: Float
SharingMode: Shared
PerformanceMode: LowLatency
CurrentState: Open
XRunCount: 0
FramesRead: 0
FramesWritten: 0
openal-log(1.405184s): [ALSOFT] (II) Post-reset: Stereo, Float32, 48000hz, 480 / 1920 buffer
openal-log(1.405338s): [ALSOFT] (II) Stereo rendering
openal-log(1.406232s): [ALSOFT] (II) Channel config, Main: 3, Real: 2
openal-log(1.406254s): [ALSOFT] (II) Allocating 5 channels, 20480 bytes
openal-log(1.406286s): [ALSOFT] (II) Enabling single-band first-order ambisonic decoder
openal-log(1.406316s): [ALSOFT] (II) Max sources: 256 (255 + 1), effect slots: 64, sends: 2
openal-log(1.406338s): [ALSOFT] (II) Dithering disabled
openal-log(1.406348s): [ALSOFT] (II) Output limiter disabled
openal-log(1.406356s): [ALSOFT] (II) Fixed device latency: 0ns
openal-log(1.409398s): [ALSOFT] (II) Post-start: Stereo, Float32, 48000hz, 480 / 1920 buffer
```

**Comment by kcat** [comment]
> I'm noticing that the stream properties seem to be a bit different after the reset (and more in line with what the backend seems to ask for; ie LowLatency mode). Is that meaningful?

Actually, yes. On the one hand, that makes it appear like the device is being reset at a lower level (hardware, kernel), which would cause a disconnection/temporary unavailability. Though I don't know why it would allow the stream to connect first before disconnecting. It's also odd that it will connect with different parameters (in the earlier log that has the device open and context creation fails, the stream gets created with the requested 48khz low-latency mode before failing to start, in the subsequent log the context gets created and the stream gets an alternate 44.1khz without low-latency mode, which starts before getting disconnected afterward, then the stream is recreated as the requested 48khz with low-latency). The DeviceId also changes from `9779` to `3`. Maybe some kind of hardware or driver bug causing the device to initiate a reconfiguration, while also returning success or failure before the reconfiguration completes.

Given what I see in that latest log, though, I wonder if low-latency mode may be overkill for OpenAL Soft. The documentation isn't clear about what kind of overall latency to expect when not requesting any particular latency mode, but an update size (`FramesPerBurst`) of 20ms (with "None" performance mode) is closer to what I'd expect and prefer, while an update size of 5ms (with LowLatency mode) is quite extreme. Perhaps removing the LowLatency request would stop the spurious disconnects from happening and give a more reasonable update size. But it would be a problem if none/unspecified performance mode could result in hundreds of milliseconds between when OpenAL Soft writes some audio to the stream and when it's finally heard. It would also be a problem if that would also lower CPU performance and cause more dropouts. I wish there was more information about this stuff.

**Comment by efroemling** [comment]
Do you think the disconnect behavior is just something that should be expected occasionally? Even if removing low-latency-mode fixes it in some cases, the earlier logs point out an immediate disconnect that seems to happen despite having low-latency-mode already. Should alcCreateContext perhaps succeed even if the device is currently disconnected? (presumably allowing the same reset code that is needed to handle later disconnects to handle that case too?)

**Comment by kcat** [comment]
> Do you think the immediate disconnect behavior is just something that should be expected occasionally?

This is the first I'm really seeing it. But if removing the low latency mode doesn't completely fix it (or if low latency mode can't be removed due to creating latency or performance issues), it may be something to expect with certain Android devices. In either case, I don't see much of an issue trying to work around it by continuing to retry opening the device and create a context (aside from making the code to do it), assuming it will work after a short enough time.

> Should alcCreateContext perhaps succeed even if the device is currently disconnected? (presumably allowing the same reset code that is needed to handle later disconnects to handle that case too?)

I'm not sure. Especially for apps that aren't expecting disconnects, having context creation fail would make the failure more apparent to more apps. Maybe I could have the Oboe backend watch for those errors and retry itself before returning failure, though it would depend on how long it could take.

**Comment by efroemling** [comment]
On the low-latency front, I just stumbled across this which does recommend using low latency mode in Oboe for games:
https://developer.android.com/games/sdk/oboe/low-latency-audio

It also mentions a few other things such as oboe::Usage::Game and oboe::SharingMode::Exclusive. Think it would be worth setting any of that stuff?

As far as the disconnected alcCreateContext case, I'm totally fine with the retry approach if you feel that's best. I've got a basic form of that in my current build so I'll keep waiting for logs and see if/how that works. But if you'd like me to try without low latency mode or anything else I'd be happy to push out another test build; just holler.

**Comment by kcat** [comment]
> It also mentions a few other things such as oboe::Usage::Game and oboe::SharingMode::Exclusive. Think it would be worth setting any of that stuff?

`Game` usage probably. Not sure about `Exclusive` since that would prevent anything else from using the audio device, and be prevented from using the device if something else also is at the time.

**Comment by efroemling** [comment]
Looking around a bit more [at the bottom of that page](https://developer.android.com/games/sdk/oboe/low-latency-audio#checklist_results), they note that low latency mode gives a latency of ~26ms while regular mode is ~205ms. I tried the OboeTester app they link to on a test device and got similar results. With that big of a difference I definitely feel low latency mode is worthwhile. (Just speaking for my own use case; not what'd be best for OpenAL as a whole).

Anyway; I think I'm good to go now for the most point on my implementation.
It does seem that my alcCreateContext retry code is working at least sometimes for handling these rare immediate disconnects, and for cases where it keeps failing I hacked in an option to set `builder.setAudioApi(oboe::AudioApi::OpenSLES)` in OboePlayback::reset(), which is seeming to work well as a last resort catch-all. I'm also using that to work around a particular string of crashes deep in the heart of libaudioclient.so that only occurs on Xiaomi Android 11 devices; they seem cranky with the Aaudio Oboe backend but fine with OpenSL.

Thanks again for all the help and feedback on this stuff.
